File: px-checkout/src/main/java/com/mercadopago/android/px/model/internal/CheckoutFeatures.java
package com.mercadopago.android.px.model.internal;

import androidx.annotation.NonNull;
import com.google.gson.annotations.SerializedName;
import java.util.ArrayList;
import java.util.List;

/**
 * Checkout features contains feature specific params and metadata about integration.
 */
public final class CheckoutFeatures {

    @SerializedName("one_tap")
    private final boolean express;
    private final boolean split;
    @SerializedName("odr")
    private final boolean odrFlag;
    private final boolean comboCard;
    private final boolean hybridCard;
    private final List<String> validationPrograms;

    /* default */ CheckoutFeatures(@NonNull final Builder builder) {
        express = builder.express;
        split = builder.split;
        odrFlag = builder.odrFlag;
        comboCard = builder.comboCard;
        hybridCard = builder.hybridCard;
        validationPrograms = builder.validationPrograms;
    }

    public static final class Builder {

        /* default */ boolean split;
        /* default */ boolean express;
        /* default */ boolean odrFlag;
        /* default */ boolean comboCard;
        /* default */ boolean hybridCard;
        /* default */ List<String> validationPrograms = new ArrayList<>();

        public Builder setSplit(final boolean split) {
            this.split = split;
            return this;
        }

        public Builder setExpress(final boolean express) {
            this.express = express;
            return this;
        }

        public Builder setOdrFlag(final boolean odrFlag) {
            this.odrFlag = odrFlag;
            return this;
        }

        public Builder setComboCard(final boolean comboCard) {
            this.comboCard = comboCard;
            return this;
        }

        public Builder setHybridCard(final boolean hybridCard) {
            this.hybridCard = hybridCard;
            return this;
        }

        public Builder addValidationPrograms(final List<String> validationPrograms) {
            this.validationPrograms.addAll(validationPrograms);
            return this;
        }

        public CheckoutFeatures build() {
            return new CheckoutFeatures(this);
        }
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/model/internal/CheckoutFeatures.kt
package com.mercadopago.android.px.model.internal

import com.google.gson.annotations.SerializedName

/**
 * Checkout features contains feature specific params and metadata about integration.
 */
internal class CheckoutFeatures(builder: Builder) {

    @SerializedName("one_tap")
    val express: Boolean
    val split: Boolean

    @SerializedName("odr")
    val odrFlag: Boolean
    val comboCard: Boolean
    val hybridCard: Boolean
    val pix: Boolean
    val customTaxesCharges: Boolean

    @SerializedName("validations_programs")
    val validationPrograms: List<String>

    /* default */
    init {
        express = builder.express
        split = builder.split
        odrFlag = builder.odrFlag
        comboCard = builder.comboCard
        hybridCard = builder.hybridCard
        pix = builder.pix
        customTaxesCharges = builder.customTaxesCharges
        validationPrograms = builder.validationPrograms
    }

    class Builder {
        var split = false
        var express = false
        var odrFlag = false
        var comboCard = false
        var hybridCard = false
        var pix = false
        var customTaxesCharges = false
        var validationPrograms: MutableList<String> = mutableListOf()

        fun setSplit(split: Boolean) = apply { this.split = split }

        fun setExpress(express: Boolean) = apply { this.express = express }

        fun setOdrFlag(odrFlag: Boolean) = apply { this.odrFlag = odrFlag }

        fun setComboCard(comboCard: Boolean) = apply { this.comboCard = comboCard }

        fun setHybridCard(hybridCard: Boolean) = apply { this.hybridCard = hybridCard }

        fun setPix(pix: Boolean) = apply { this.pix = pix }

        fun setCustomTaxesCharges(customTaxesCharges: Boolean) = apply { this.customTaxesCharges = customTaxesCharges }

        fun addValidationPrograms(validationPrograms: List<String>) =
            apply { this.validationPrograms.addAll(validationPrograms) }

        fun addValidationProgram(validationProgram: String) = apply { validationPrograms.add(validationProgram) }

        fun build() = CheckoutFeatures(this)
    }
}
